src/rebar_prv_common_test.erl

```diff
@@ -37,54 +38,352 @@ init(State) ->
+    end.
+
+join(undefined, Suites) -> Suites;
+join(Dir, Suites) when is_list(Dir), is_integer(hd(Dir)) ->
+    lists:map(fun(S) -> filename:join([Dir, S]) end, Suites);
+%% multiple dirs or a bad dir argument, try to continue anyways
+join(_, Suites) -> Suites. 
+
+find_suite_dirs(Suites) ->
+    AllDirs = lists:map(fun(S) -> filename:dirname(filename:absname(S)) end, Suites),
+    %% eliminate duplicates
+    lists:usort(AllDirs).
+
+copy(State, Target) ->
+    case retarget_path(State, Target) of
+        %% directory lies outside of our project's file structure so
+        %%  don't copy it
+        Target    -> Target;
+        NewTarget ->
+            %% unlink the directory if it's a symlink
+            case ec_file:is_symlink(NewTarget) of
+                true  -> ok = ec_file:remove(NewTarget);
+                false -> ok
+            end,
+            ok = ec_file:copy(Target, NewTarget, [recursive]),
+            NewTarget
+    end.
+
+compile_dir(State, Dir) ->
+    NewState = replace_src_dirs(State, [Dir]),
+    ok = rebar_erlc_compiler:compile(NewState, rebar_dir:base_dir(State), Dir),
+    ok = maybe_cover_compile(State, Dir),
+    Dir.
+
+retarget_path(State, Path) ->
+    ProjectApps = rebar_state:project_apps(State),
+    retarget_path(State, Path, ProjectApps).
+
+%% not relative to any apps in project, check to see it's relative to
+%%  project root
+retarget_path(State, Path, []) ->
+    case relative_path(reduce_path(Path), rebar_state:dir(State)) of
+        {ok, NewPath}         -> filename:join([rebar_dir:base_dir(State), NewPath]);
+        %% not relative to project root, don't modify
+        {error, not_relative} -> Path
+    end;
+%% relative to current app, retarget to the same dir relative to
+%%  the app's out_dir
+retarget_path(State, Path, [App|Rest]) ->
+    case relative_path(reduce_path(Path), rebar_app_info:dir(App)) of
+        {ok, NewPath}         -> filename:join([rebar_app_info:out_dir(App), NewPath]);
+        {error, not_relative} -> retarget_path(State, Path, Rest)
+    end.
+
+relative_path(Target, To) ->
+    relative_path1(filename:split(filename:absname(Target)),
+                   filename:split(filename:absname(To))).
+
+relative_path1([Part|Target], [Part|To]) -> relative_path1(Target, To);
+relative_path1([], [])                   -> {ok, ""};
+relative_path1(Target, [])               -> {ok, filename:join(Target)};
+relative_path1(_, _)                     -> {error, not_relative}.
+
+reduce_path(Dir) -> reduce_path([], filename:split(filename:absname(Dir))).
+
+reduce_path([], [])                -> filename:nativename("/");
+reduce_path(Acc, [])               -> filename:join(lists:reverse(Acc));
+reduce_path(Acc, ["."|Rest])       -> reduce_path(Acc, Rest);
+reduce_path([_|Acc], [".."|Rest])  -> reduce_path(Acc, Rest);
+reduce_path([], [".."|Rest])       -> reduce_path([], Rest);
+reduce_path(Acc, [Component|Rest]) -> reduce_path([Component|Acc], Rest).
```

Comment:
I'll make a mental note to remember all of that stuff in case it's needed again to be extracted into `rebar_dir` or `rebar_utils`